common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchLogStore.scala

```
@@ -0,0 +1,138 @@
+  private val esClient = new ElasticSearchRestClient(protocol, host, port, httpFlow)
+
+  private def transcribeLogs(queryResult: EsSearchResult): ActivationLogs = {
+    val logs = queryResult.hits.hits.map(hit => {
+      val userLogEntry = hit.source.convertTo[UserLogEntry]
+      s"${userLogEntry.time} ${userLogEntry.stream}: ${userLogEntry.message.stripLineEnd}"
+    })
+
+    ActivationLogs(logs)
+  }
+
+  private def extractRequiredHeaders(request: HttpRequest) =
+    request.headers.filter {
+      case header: HttpHeader if requiredHeaders.contains(header.lowercaseName) => true
+      case _                                                                    => false
+    }.toList
+
+  private def generatePayload(activation: WhiskActivation) = {
+    val logQuery = s"_type: $logMessageField AND $activationIdField: ${activation.activationId}"
+    val queryString = EsQueryString(EsQueryValue(logQuery))
+    val sort = Some(Array(EsQueryTimestamp(EsQueryOrder("asc")).toJson))
+
+    EsQuery(queryString.toJson, sort).toJson.asJsObject
+  }
+
+  private def generatePath(user: Identity) = {
+    val date = LocalDate.now.format(DateTimeFormatter.ofPattern("yyyy.MM.dd"))
+
+    Uri(s"/elasticsearch/logstash-${user.uuid.asString}-$date/_search")
+  }
+
+  override def fetchLogs(user: Identity, activation: WhiskActivation, request: HttpRequest): Future[ActivationLogs] = {
+    val headers = extractRequiredHeaders(request)
+
+    // Return logs from ElasticSearch, or return logs from activation if required headers are not present
+    if (headers.length == requiredHeaders.length) {
+      esClient.post(generatePath(user), headers, generatePayload(activation)).map { response =>
+        response match {
+          case Right(queryResult) =>
+            transcribeLogs(queryResult.convertTo[EsSearchResult])
+          case Left(code) =>
+            throw new RuntimeException(s"Status code '$code' was returned from log store")
```

This error is not surfaced to the user.